nix flake for liquidhaskell & component packages

plredmond opened the pull request:
This is the nix flake I maintain [over here](https://github.com/ucsd-progsys/liquidhaskell/compare/develop...plredmond:nix-flake#files_bucket), with all the history cut out.

Nix flakes allow nix users to easily build and depend on other projects, without integrating the build instructions into nixpkgs. It's like a project-format for nix.

This one is admittedly complex, and so I think it needs some review and improvement by other LH users who are nix users.

TODO:

* [ ] Check for references to the deleted nix files (eg. there's one in shell.yaml) and create appropriate replacements with [flake-compat](https://github.com/edolstra/flake-compat); make sure to match the devtools and various usages supported by the deleted nix files.
* [ ] Update the INSTALL.md nix instructions as suggested by @tbidne

plredmond committed:
nix flake from https://github.com/ucsd-progsys/liquidhaskell/compare/develop...plredmond:nix-flake (but without the history)

plredmond commented:
Separate PR for liquid-fixpoint nix flake: https://github.com/ucsd-progsys/liquid-fixpoint/pull/582

plredmond commented on flake.lock:
```diff
@@ -0,0 +1,68 @@
+      "locked": {
+        "lastModified": 1653693908,
+        "narHash": "sha256-IYrQB9M/XdDaDzQ1576iUNvbehJ4fUlKvPtg5l/Z5xU=",
+        "owner": "plredmond",
```
Fix this to point at the ucsd-progsys/liquid-fixpoint if/when https://github.com/ucsd-progsys/liquid-fixpoint/pull/582 is merged.

plredmond commented on flake.nix:
```diff
@@ -0,0 +1,131 @@
+    nixpkgs.url = github:NixOS/nixpkgs/nixos-22.05;
+    flake-utils.url = github:numtide/flake-utils;
+
+    liquid-fixpoint.url = github:plredmond/liquid-fixpoint/nix-flake; # TODO change to official repo after merge
```
Fix this to point at the ucsd-progsys/liquid-fixpoint if/when https://github.com/ucsd-progsys/liquid-fixpoint/pull/582 is merged.

tbidne commented:
In general, I think this is a great idea. I have had trouble getting tests to consistently pass locally, so anything that can make that smoother would be greatly appreciated. I have some scattered thoughts:

* The nix files you deleted are referenced in the [stack.yaml](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/stack.yaml#L35). If the flake here obviates that usage, then they should be removed. However, it's worth asking if there are nix-users who would still want that workflow (e.g. non-flake nix users).
* [INSTALL.md](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/INSTALL.md) should be updated with the new nix instructions.
* I noticed the `devShell` loads with `ghc` but no `cabal` nor `stack`. The deleted [shell.nix](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/shell.nix) included the latter, so you probably want to as well (e.g. support incremental compilation).
* Re: running tests locally: Does `nix build .#liquidhaskell_with_tests` pass for you? I encounter some `synthesis` errors in the test suite. Admittedly this is not directly related to your PR, but as easy local testing is my primary motivator, I am wondering if I am doing something wrong.

Thanks!

plredmond commented:
Thanks for your feedback @tbidne. Responses below.

#### Tests

> I have had trouble getting tests to consistently pass locally, so anything that can make that smoother would be greatly appreciated.
> * Re: running tests locally: Does nix build .#liquidhaskell_with_tests pass for you? I encounter some synthesis errors in the test suite. Admittedly this is not directly related to your PR, but as easy local testing is my primary motivator, I am wondering if I am doing something wrong.

When building this nix-flake there are synthesis tests and syntax tests running (and this matches what's in the [LH cabalfile](https://github.com/ucsd-progsys/liquidhaskell/blob/develop/liquidhaskell.cabal) currently). The synthesis tests fail for me too (yay reproducibility). Should there be more tests running? Do we expect the synthesis tests to pass?

#### Deleted nix files

> * The nix files you deleted are referenced in the [stack.yaml](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/stack.yaml#L35). If the flake here obviates that usage, then they should be removed. However, it's worth asking if there are nix-users who would still want that workflow (e.g. non-flake nix users).
> * I noticed the devShell loads with ghc but no cabal nor stack. The deleted [shell.nix](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/shell.nix) included the latter, so you probably want to as well (e.g. support incremental compilation).

I've added a TODO to the PR description to create replacements using [flake-compat](https://github.com/edolstra/flake-compat). I think that's the expected best practice for the time being.

#### Documentation

> * [INSTALL.md](https://github.com/ucsd-progsys/liquidhaskell/blob/d9e2e4b063b4facc605ccd939cb5f27aec4ce1a6/INSTALL.md) should be updated with the new nix instructions.

Added as a TODO to the PR description.

ranjitjhala commented:
Lets merge this @facundominguez ?

facundominguez commented:
I'm merging it in #2066. Note, though, that I'm not myself a user of it.

plredmond commented:
Ah, well, I haven't had time yet to do the TODOs above. I'll make an issue so that it doesn't get lost.

facundominguez commented:
> Check for references to the deleted nix files

I took care of this one (I think), and for now I removed all references to nix in the README and the INSTALL file.
